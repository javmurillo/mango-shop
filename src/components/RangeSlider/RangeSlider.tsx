import React, { Component } from 'react';
import Track from '../Track/Track';
import RangeBullet from '../RangeBullet/RangeBullet';
import styled from 'styled-components';
import { RangeSliderProps } from './range-slider.props.interface';
import { RangeSliderState } from './range-slider.state.interface';
import { parseConfigFileTextToJson } from 'typescript';

const StyledRangeSlider = styled.div`
  position: relative;
  width: 100%;
  height: 34px;
`;

export default class RangeSlider extends Component<
  RangeSliderProps,
  RangeSliderState
> {
  public state: RangeSliderState;
  public factor = 1;
  public step: number | number[];
  public startSteps: {
    left: number;
    right: number;
  };
  public endSteps: {
    left: number;
    right: number;
  };
  public startIndex?: number;
  public endIndex?: number;

  constructor(props: RangeSliderProps) {
    super(props);
    const start = props.rangeValue.start;
    const end = props.rangeValue.end;
    const step = props.step;
    this.step = step;

    this.state = {
      start,
      end,
    };
    if (Array.isArray(step)) {
      this.startIndex = 0;
      this.endIndex = step.length - 1;
      this.startSteps = {
        left: 0,
        right: step[this.startIndex! + 1] - step[this.startIndex!],
      };
      this.endSteps = {
        left: step[this.endIndex!] - step[this.endIndex! - 1],
        right: 0,
      };
    } else {
      this.startSteps = {
        left: step,
        right: step,
      };
      this.endSteps = {
        left: step,
        right: step,
      };
    }
  }

  static getDerivedStateFromProps(
    props: RangeSliderProps,
    state: RangeSliderState
  ): {
    start: number;
    end: number;
  } | null {
    if (
      props.rangeValue.start !== state.start ||
      props.rangeValue.end !== state.end
    ) {
      return {
        start: props.rangeValue && props.rangeValue.start,
        end: props.rangeValue && props.rangeValue.end,
      };
    }
    return null;
  }

  setTrackDimensions = (track: HTMLDivElement): void => {
    if (track) {
      const trackLength = track.clientWidth;
      this.setState({
        trackLength,
      });
    }
  };

  setHandleSize = (bullet: HTMLDivElement | null): void => {
    if (bullet) {
      const handleSize = bullet.clientWidth;
      if (!this.state.handleSize) {
        this.setState({
          handleSize,
        });
      }
    }
  };

  updateIndexes(increase: number, key: 'start' | 'end'): void {
    if (key === 'start') {
      increase === 1 ? this.startIndex!++ : this.startIndex!--;
    } else if (key === 'end') {
      increase === 1 ? this.endIndex!++ : this.endIndex!--;
    }
  }

  updateSteps(step: number[], key: 'start' | 'end'): void {
    if (key === 'start') {
      this.startSteps = {
        left: step[this.startIndex!] - step[this.startIndex! - 1],
        right: step[this.startIndex! + 1] - step[this.startIndex!],
      };
    } else if (key === 'end') {
      this.endSteps = {
        left: step[this.endIndex!] - step[this.endIndex! - 1],
        right: step[this.endIndex! + 1] - step[this.endIndex!],
      };
    }
  }

  getStartStep(increase: number): number {
    if (Array.isArray(this.step)) {
      return increase === 1 ? this.startSteps?.right : this.startSteps?.left;
    }
    return this.step;
  }

  getEndStep(increase: number): number | undefined {
    if (Array.isArray(this.step)) {
      return increase === 1 ? this.endSteps?.right : this.endSteps?.left;
    }
    return this.step;
  }

  startHandleMove = (increase: number): void => {
    const { start } = this.state;
    const calculatedStep = this.getStartStep(increase);
    if (Array.isArray(this.step)) {
      this.updateIndexes(increase, 'start');
      this.updateSteps(this.step as number[], 'start');
    }
    const nextStart = start + increase * calculatedStep!;
    const updatedStart = this.getStartValue(nextStart);
    if (updatedStart !== start) {
      this.updateState(updatedStart, this.state.end);
      this.onChange(updatedStart, this.state.end);
    }
  };

  endHandleMove = (increase: number): void => {
    const { end } = this.state;
    const calculatedStep = this.getEndStep(increase);
    if (Array.isArray(this.step)) {
      this.updateIndexes(increase, 'end');
      this.updateSteps(this.step as number[], 'end');
    }

    const nextEnd = end + increase * calculatedStep!;
    const updatedEnd = this.getEndValue(nextEnd);
    if (updatedEnd !== end) {
      this.updateState(this.state.start, updatedEnd);
      this.onChange(this.state.start, updatedEnd);
    }
  };

  getStartValue(nextStart: number): number {
    if (nextStart < this.props.min) {
      return this.props.min;
    } else if (nextStart > this.state.end) {
      return this.state.end;
    }
    return nextStart;
  }

  getEndValue(nextEnd: number): number {
    if (nextEnd > this.props.max) {
      return this.props.max;
    } else if (nextEnd < this.state.start) {
      return this.state.start;
    }
    return nextEnd;
  }

  updateState = (start: number, end: number): void => {
    this.setState({
      start,
      end,
    });
  };

  onChange = (start: number, end: number): void => {
    this.props.onChange({
      start,
      end,
    });
  };

  render(): Object {
    let startValue = 0;
    let endValue = 0;
    let percentageFactor = 1;
    const { handleSize, trackLength, start, end } = this.state;
    const { min, max } = this.props;
    if (trackLength && handleSize) {
      const calculatedTrackWidth = trackLength - handleSize;
      this.factor = calculatedTrackWidth / (max - min);
      startValue = (start - min) * this.factor;
      endValue = (end - min) * this.factor;
      percentageFactor = 100 / trackLength;
    }
    return (
      <StyledRangeSlider>
        <Track trackRef={this.setTrackDimensions} />
        <RangeBullet
          offset={`${startValue * percentageFactor}%`}
          handleRef={this.setHandleSize}
          handleMove={this.startHandleMove}
          factor={this.factor}
          step={this.startSteps}
        />
        <RangeBullet
          offset={`${endValue * percentageFactor}%`}
          handleRef={this.setHandleSize}
          handleMove={this.endHandleMove}
          factor={this.factor}
          step={this.endSteps}
        />
      </StyledRangeSlider>
    );
  }
}
