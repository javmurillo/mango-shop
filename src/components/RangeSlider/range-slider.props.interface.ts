export interface RangeSliderProps {
  min: number;
  max: number;
  step: number;
  rangeValue: {
    start: number;
    end: number;
  };
  onChange?: Function;
}
