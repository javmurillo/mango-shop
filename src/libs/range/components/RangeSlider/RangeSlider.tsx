import React, { Component } from 'react';
import Track from '../RangeTrack/RangeTrack';
import RangeBullet from '../RangeBullet/RangeBullet';
import styled from 'styled-components';
import { RangeSliderProps } from './range-slider.props.interface';
import { RangeSliderState } from './range-slider.state.interface';

const StyledRangeSlider = styled.div`
  position: relative;
  width: 100%;
  height: 34px;
`;

export default class RangeSlider extends Component<
  RangeSliderProps,
  RangeSliderState
> {
  public state: RangeSliderState;
  public factor = 1;
  public startSteps: {
    left: number;
    right: number;
  };
  public endSteps: {
    left: number;
    right: number;
  };
  public startIndex?: number;
  public endIndex?: number;

  constructor(props: RangeSliderProps) {
    super(props);
    const start = props.rangeValue.start;
    const end = props.rangeValue.end;
    const step = props.step;
    this.state = {
      start,
      end,
    };
    let startLeftStep = step;
    let startRightStep = step;
    let endLeftStep = step;
    let endRightStep = step;
    if (Array.isArray(step)) {
      this.startIndex = 0;
      this.endIndex = step.length - 1;
      startLeftStep = 0;
      startRightStep = step[this.startIndex! + 1] - step[this.startIndex!];
      endLeftStep = step[this.endIndex!] - step[this.endIndex! - 1];
      endRightStep = 0;
    }
    this.startSteps = {
      left: startLeftStep as number,
      right: startRightStep as number,
    };
    this.endSteps = {
      left: endLeftStep as number,
      right: endRightStep as number,
    };
  }

  static getDerivedStateFromProps(
    props: RangeSliderProps,
    state: RangeSliderState
  ): {
    start: number;
    end: number;
  } | null {
    if (
      props.rangeValue.start !== state.start ||
      props.rangeValue.end !== state.end
    ) {
      return {
        start: props.rangeValue && props.rangeValue.start,
        end: props.rangeValue && props.rangeValue.end,
      };
    }
    return null;
  }

  setTrackDimensions = (track: HTMLDivElement): void => {
    if (track) {
      const trackLength = track.clientWidth;
      this.setState({
        trackLength,
      });
    }
  };

  setHandleSize = (bullet: HTMLDivElement | null): void => {
    if (bullet) {
      const handleSize = bullet.clientWidth;
      if (!this.state.handleSize) {
        this.setState({
          handleSize,
        });
      }
    }
  };

  updateIndexes(increase: number, step: number[], key: 'start' | 'end'): void {
    if (key === 'start') {
      if (increase === 1 && this.startIndex! < step.length - 1) {
        this.startIndex!++;
      } else if (increase === -1 && this.startIndex! > 0) {
        this.startIndex!--;
      }
    } else if (key === 'end') {
      if (increase === 1 && this.endIndex! < step.length - 1) {
        this.endIndex!++;
      } else if (increase === -1 && this.endIndex! > 0) {
        this.endIndex!--;
      }
    }
  }

  updateSteps(step: number[], key: 'start' | 'end'): void {
    if (key === 'start') {
      const startLeftStep = step[this.startIndex!] - step[this.startIndex! - 1];
      const startRightStep =
        step[this.startIndex! + 1] - step[this.startIndex!];
      this.startSteps = {
        left: !isNaN(startLeftStep) ? startLeftStep : 0,
        right: !isNaN(startRightStep) ? startRightStep : 0,
      };
    } else if (key === 'end') {
      const endRightStep = step[this.endIndex! + 1] - step[this.endIndex!];
      const endLeftStep = step[this.endIndex!] - step[this.endIndex! - 1];
      this.endSteps = {
        left: !isNaN(endLeftStep) ? endLeftStep : 0,
        right: !isNaN(endRightStep) ? endRightStep : 0,
      };
    }
  }

  getStartStep(
    increase: number,
    step: RangeSliderProps['step'],
    startSteps: {
      left: number;
      right: number;
    }
  ): number {
    if (Array.isArray(step)) {
      return increase === 1 ? startSteps.right : startSteps.left;
    }
    return step;
  }

  getEndStep(
    increase: number,
    step: RangeSliderProps['step'],
    endSteps: {
      left: number;
      right: number;
    }
  ): number {
    if (Array.isArray(step)) {
      return increase === 1 ? endSteps.right : endSteps.left;
    }
    return step;
  }

  startHandleMove = (increase: number): void => {
    const { start } = this.state;
    const calculatedStep = this.getStartStep(
      increase,
      this.props.step,
      this.startSteps
    );
    if (Array.isArray(this.props.step)) {
      this.updateIndexes(increase, this.props.step, 'start');
      this.updateSteps(this.props.step as number[], 'start');
    }
    const nextStart = start + increase * calculatedStep!;
    const updatedStart = this.getStartValue(
      nextStart,
      this.props.min,
      this.state.end
    );
    if (updatedStart !== start) {
      this.updateState(updatedStart, this.state.end);
      this.onChange(updatedStart, this.state.end);
    }
  };

  endHandleMove = (increase: number): void => {
    const { end } = this.state;
    const calculatedStep = this.getEndStep(
      increase,
      this.props.step,
      this.endSteps
    );
    if (Array.isArray(this.props.step)) {
      this.updateIndexes(increase, this.props.step, 'end');
      this.updateSteps(this.props.step as number[], 'end');
    }
    const nextEnd = end + increase * calculatedStep!;
    const updatedEnd = this.getEndValue(
      nextEnd,
      this.props.max,
      this.state.start
    );
    if (updatedEnd !== end) {
      this.updateState(this.state.start, updatedEnd);
      this.onChange(this.state.start, updatedEnd);
    }
  };

  getStartValue(nextStart: number, min: number, end: number): number {
    if (nextStart < min) {
      return min;
    } else if (nextStart > end) {
      return end;
    }
    return nextStart;
  }

  getEndValue(nextEnd: number, max: number, start: number): number {
    if (nextEnd > max) {
      return max;
    } else if (nextEnd < start) {
      return start;
    }
    return nextEnd;
  }

  updateState = (start: number, end: number): void => {
    this.setState({
      start,
      end,
    });
  };

  onChange = (start: number, end: number): void => {
    const parsedStart = Number(start.toFixed(2));
    const parsedEnd = Number(end.toFixed(2));
    this.props.onChange({
      start: parsedStart,
      end: parsedEnd,
    });
  };

  render(): Object {
    let startValue = 0;
    let endValue = 0;
    let percentageFactor = 1;
    const { handleSize, trackLength, start, end } = this.state;
    const { min, max } = this.props;
    if (trackLength && handleSize) {
      const calculatedTrackWidth = trackLength - handleSize;
      this.factor = calculatedTrackWidth / (max - min);
      startValue = (start - min) * this.factor;
      endValue = (end - min) * this.factor;
      percentageFactor = 100 / trackLength;
    }
    return (
      <StyledRangeSlider>
        <Track trackRef={this.setTrackDimensions} />
        <RangeBullet
          offset={`${startValue * percentageFactor}%`}
          handleRef={this.setHandleSize}
          handleMove={this.startHandleMove}
          factor={this.factor}
          step={this.startSteps}
        />
        <RangeBullet
          offset={`${endValue * percentageFactor}%`}
          handleRef={this.setHandleSize}
          handleMove={this.endHandleMove}
          factor={this.factor}
          step={this.endSteps}
        />
      </StyledRangeSlider>
    );
  }
}
