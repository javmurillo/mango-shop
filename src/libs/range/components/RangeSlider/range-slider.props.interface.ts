export interface RangeSliderProps {
  min: number;
  max: number;
  step: number | number[];
  rangeValue: {
    start: number;
    end: number;
  };
  onChange: (rangeValue: { start: number; end: number }) => void;
}
