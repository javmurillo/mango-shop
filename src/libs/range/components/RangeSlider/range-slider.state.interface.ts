import { RangeState } from '../Range/range-state.interface';

export interface RangeSliderState {
  start: RangeState['rangeValue']['start'];
  end: RangeState['rangeValue']['end'];
  handleSize?: number;
  trackLength?: number;
}
