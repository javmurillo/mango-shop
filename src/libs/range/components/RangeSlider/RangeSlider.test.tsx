import { cleanup, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import TestRenderer from 'react-test-renderer';
import RangeSlider from './RangeSlider';

afterEach(cleanup);

const rangeValue = { start: 0, end: 100 };

const setup = () => {
  const utils = render(
    <RangeSlider
      min={0}
      max={100}
      step={5}
      rangeValue={rangeValue}
      ariaLabel="range-slider"
    />
  );
  const rangeSlider = utils.getByLabelText('range-slider');
  return {
    rangeSlider,
    ...utils,
  };
};

describe('<RangeSlider/> tests', () => {
  test('renders correctly', () => {
    const { rangeSlider } = setup();

    expect(rangeSlider).toBeInTheDocument();
  });

  test('state initialized correctly', () => {
    const { rangeSlider } = setup();

    expect(rangeSlider).toBeInTheDocument();
  });

  test('shoud match snapshot', () => {
    const testRendererJson = TestRenderer.create(
      <RangeSlider
        min={0}
        max={100}
        step={5}
        rangeValue={rangeValue}
        ariaLabel="range-slider"
      />
    ).toJSON();
    expect(testRendererJson).toMatchSnapshot();
  });
});
