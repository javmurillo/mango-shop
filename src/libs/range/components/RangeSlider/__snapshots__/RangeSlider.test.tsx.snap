// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RangeSlider/> tests shoud match snapshot 1`] = `
<div
  aria-label="range-slider"
  className="sc-gsTCUz ipuGXO"
>
  <div
    className="sc-bdfBwQ cIBWvr"
  />
  <div
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "backgroundColor": "black",
        "border": "1px solid #052350",
        "borderRadius": "50%",
        "cursor": "grab",
        "display": "inline-block",
        "height": 30,
        "left": "0%",
        "position": "absolute",
        "width": 30,
      }
    }
  />
  <div
    onMouseDown={[Function]}
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    style={
      Object {
        "backgroundColor": "black",
        "border": "1px solid #052350",
        "borderRadius": "50%",
        "cursor": "grab",
        "display": "inline-block",
        "height": 30,
        "left": "0%",
        "position": "absolute",
        "width": 30,
      }
    }
  />
</div>
`;
